### PR TITLE
Introduce edit collector

### PR DESCRIPTION
**Pull request by Connor1996**
Introduce edit collector to apply the VersionEdits in batch

### CODE REVIEW

File: src/version_edit.h [hunk 1]
```diff
@@ -42,5 +45,52 @@ class VersionEdit {
...
+      collector.DeleteFile(file.first, file.second);
+      if (!s.ok()) return s;
+    }
+    return Status::OK();
+  }
+
+ private:
+  friend class VersionSet;
+
+  class Collector {
```
Comment (yiwu-arbug) [inline src/version_edit.h]: s/Collector/CFEditCollector ?

File: src/version_edit.h [hunk 2]
```diff
@@ -42,5 +45,52 @@ class VersionEdit {
...
+   public:
+    Collector() {}
```
Comment (yiwu-arbug) [inline src/version_edit.h]: nit: this can be remove.

File: src/version_edit.h [hunk 3]
```diff
@@ -42,5 +45,52 @@ class VersionEdit {
...
+
+    Status AddFile(const std::shared_ptr<BlobFileMeta>& file) {
+      auto number = file->file_number();
+      if (added_files_.find(number) != added_files_.end()) {
```
Comment (yiwu-arbug) [inline src/version_edit.h]: nit: could be just `added_files_.count(number) > 0`.

File: src/version_edit.h [hunk 4]
```diff
@@ -42,5 +45,52 @@ class VersionEdit {
...
+        return Status::Corruption("blob file has been added before");
+      }
+      added_files_.emplace(number, file);
+      return Status::OK();
+    }
+
+    Status DeleteFile(uint64_t number, SequenceNumber obsolete_sequence) {
+      if (!added_files_.erase(number)) {
+        deleted_files_.emplace_back(std::make_pair(number, obsolete_sequence));
```
Comment (yiwu-arbug) [inline src/version_edit.h]: I'm a bit paranoid, but let's also check if the file is deleted before.

File: src/version_edit.h [hunk 5]
```diff
@@ -42,5 +45,52 @@ class VersionEdit {
...
+      auto number = file->file_number();
+      if (added_files_.find(number) != added_files_.end()) {
+        return Status::Corruption("blob file has been added before");
+      }
+      added_files_.emplace(number, file);
+      return Status::OK();
+    }
+
+    Status DeleteFile(uint64_t number, SequenceNumber obsolete_sequence) {
+      if (!added_files_.erase(number)) {
```
Comment (yiwu-arbug) [inline src/version_edit.h]: let's collect the set of file added and deleted, and sanity check the file don't appear later on.

File: src/version_edit.h [hunk 6]
```diff
@@ -33,6 +35,7 @@ class VersionEdit {
 
  private:
   friend class VersionSet;
+  friend class EditCollector;
 
   bool has_next_file_number_{false};
   uint64_t next_file_number_{0};
```
Comment (yiwu-arbug) [inline src/version_edit.h]: I think the logic is cleaner if `EditCollector` also handle `next_file_number_`.

File: src/version_set.cc [hunk 7]
```diff
@@ -204,52 +207,61 @@ Status VersionSet::LogAndApply(VersionEdit* edit) {
...
-    auto number = file.first;
-    auto blob_it = files.find(number);
-    if (blob_it == files.end()) {
-      fprintf(stderr, "blob file %" PRIu64 " doesn't exist before\n", number);
-      abort();
-    } else if (blob_it->second->is_obsolete()) {
-      fprintf(stderr, "blob file %" PRIu64 " has been deleted before\n",
-              number);
-      abort();
+Status VersionSet::Apply(EditCollector* collector) {
```
Comment (yiwu-arbug) [inline src/version_set.cc]: Change this to `EditCollector&`? RocksDB convention is to use reference for input param, and pointer for output param.

File: src/version_edit.h [hunk 8]
```diff
@@ -33,6 +35,7 @@ class VersionEdit {
 
  private:
   friend class VersionSet;
+  friend class EditCollector;
```
Comment (yiwu-arbug) [inline src/version_edit.h]: Adding some getters is better than make classes friend each other.

File: src/version_set.cc [hunk 9]
```diff
@@ -204,52 +207,61 @@ Status VersionSet::LogAndApply(VersionEdit* edit) {
...
       }
-      abort();
+      it->second->AddBlobFile(file.second);
     }
-    it->second->AddBlobFile(file);
-  }
 
-  it->second->ComputeGCScore();
+    it->second->ComputeGCScore();
+  }
```
Comment (yiwu-arbug) [inline src/version_set.cc]: let's also sanity check those files deleted after added in `EditCollector` does not exist in `VersionSet`.

File: src/version_edit.h (hunk 6 above)
Comment (Connor1996) [reply src/version_edit.h]: What do you mean? `EditCollector` checks whether `next_file_number_` is monotonic increasing?

Comment (Connor1996) [reply src/version_edit.h]: Oh, I got it.

File: src/edit_collector.h [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+      s = collector.AddFile(file);
+      if (!s.ok()) return s;
+    }
+    for (auto& file : edit.deleted_files_) {
+      collector.DeleteFile(file.first, file.second);
+      if (!s.ok()) return s;
+    }
+
+    if (edit.has_next_file_number_) {
+      assert(edit.next_file_number_ >= next_file_number_);
```
Comment (yiwu-arbug) [inline src/edit_collector.h]: return Corruption if this check fail?

**Pull request merged**